hibernate-core/src/main/java/org/hibernate/boot/internal/InFlightMetadataCollectorImpl.java

```diff
@@ -1847,10 +1848,12 @@ public MetadataBuildingContext getBuildingContext() {
 								}
 							}
 					);
-
-					fk.setName( nameIdentifier.render( getDatabase().getJdbcEnvironment().getDialect() ) );
+				} else {
```

Comment:
As far as how to build the Identifier... ideally `ForeignKey` would carry the Identifier, like the other `org.hibernate.mapping` objects do.  The thing you'd have to do there is to make sure that `org.hibernate.mapping.Constraint` (the `ForeignKey` super) has access to the `org.hibernate.boot.spi.MetadataBuildingContext` the way other `org.hibernate.mapping` objects that need access to services do.

Short of that, you could rely on the `fk.getName()` call already being quoted using back-ticks or JPA's double-quotes.  The `Identifier.toIdentifier( fk.getName() )` call (no boolean) wold handle that case.  You could also hook in auto-quoting checks, if enabled.